src/transformStateWithClones.js

```diff
@@ -7,7 +7,37 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const statesArray = [];
+
+  let currentState = { ...state };
+
+  actions.forEach((action) => {
+    switch (action.type) {
+      case 'clear':
+        currentState = {};
+        break;
+
+      case 'addProperties':
+        currentState = { ...currentState, ...action.extraData };
+        break;
+
+      case 'removeProperties':
+        const copy = { ...currentState };
+        action.keysToRemove.forEach((key) => {
+          delete copy[key];
+        });
+
+        currentState = { ...copy };
```

Comment:
Creating a copy of `currentState` before removing properties is unnecessary since you're already creating a new object in line 27 when deleting keys. You can directly delete keys from `currentState` and then push the modified state into `statesArray`.